algorithms/integration/kapton_2019_correction.py

```diff
@@ -0,0 +1,538 @@
+        # Get intersection between 0-2 with a-b
+        r03_ab=get_line_intersection(m03,c03,mab,cab)
+        # Get line intersection between 1-3 and a-b
+        r12_ab=get_line_intersection(m12,c12,mab,cab)
+        # Get line intersection between 0-2 and c-d
+        r03_cd=get_line_intersection(m03,c03,mcd,ccd)
+        # Get line intersection between 1-3 and c-d
+        r12_cd=get_line_intersection(m12,c12,mcd,ccd)
+        # returns pairs of x,y
+        return [(r03_ab[0], r03_ab[1], r12_ab[0], r12_ab[1]), (r03_cd[0], r03_cd[1], r12_cd[0], r12_cd[1])]
+
+class image_kapton_correction(object):
+    def __init__(self,
+                 panel_size_px = None,   #
+                 pixel_size_mm = None,   #
+                 detector_dist_mm = None,#
+                 wavelength_ang = None,
+                 reflections_sele = None,
+                 params = None,
+                 expt = None,
+                 refl = None,
+                 smart_sigmas = True,
+                 logger = None):
+        self.panel_size_px     = panel_size_px
+        self.pixel_size_mm     = pixel_size_mm
+        self.detector_dist_mm  = detector_dist_mm
+        self.wavelength_ang    = wavelength_ang
+        self.reflections_sele  = reflections_sele
+        self.params            = params
+        self.expt              = expt
+        self.refl              = refl
+        self.smart_sigmas      = smart_sigmas
+        self.logger            = logger
+        self.extract_params()
+
+    def extract_params(self):
+        h = self.params.xtal_height_above_kapton_mm.value
+        t = self.params.kapton_thickness_mm.value
+        w = self.params.kapton_half_width_mm.value
+        a = self.params.rotation_angle_deg.value
+        self.kapton_params = (h, t, w, a)
+
+        if self.smart_sigmas:
+            sig_h = self.params.xtal_height_above_kapton_mm.sigma
+            sig_t = self.params.kapton_thickness_mm.sigma
+            sig_w = self.params.kapton_half_width_mm.sigma
+            sig_a = self.params.rotation_angle_deg.sigma
+            self.kapton_params_sigmas = (sig_h, sig_t, sig_w, sig_a)
+            assert not False in [sig >=0 for sig in self.kapton_params_sigmas], "Kapton param sigmas must be nonnegative"
+            self.kapton_params_maxes = [[self.kapton_params[i] + self.kapton_params_sigmas[j]
+                                  if j==i else self.kapton_params[i]
+                                  for i in xrange(4)] for j in xrange(4)]
+            self.kapton_params_mins  = [[max(self.kapton_params[i] - self.kapton_params_sigmas[j], 0.001)
+                                  if j==i else self.kapton_params[i]
+                                  for i in xrange(3)] + [a] for j in xrange(3)] + \
+                                  [[self.kapton_params[i] for i in xrange(3)] + [a - sig_a]]
+
+    def __call__(self, plot=False):
+        def correction_and_within_spot_sigma(params_version, variance_within_spot=True):
```

Comment:
Again - why does this need to be an embedded function? It doesn't appear to capture anything except `self`, so could it just be a normal member? (with a private prefix e.g. `_correction_and_within_spot_sigma` if only intended for internal use)